pkg/remotes/docker/artifacts.go

```diff
@@ -0,0 +1,266 @@
+	switch desc.MediaType {
+	case artifactspec.MediaTypeArtifactManifest:
+		h, err := d.filterHosts(docker.HostCapabilityPush)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(h) == 0 {
+			return nil, errors.New("no host with push")
+		}
+
+		host := h[0]
+
+		err = d.CheckManifest(ctx, host, desc)
+		if err != nil {
+			return nil, err
+		}
+
+		return d.PreparePutManifest(ctx, host, desc)
+	}
+
+	pusher, err := d.Resolver.Pusher(ctx, d.reference)
+	if err != nil {
+		return nil, err
+	}
+
+	return pusher.Push(ctx, desc)
+}
+
+func (d *dockerDiscoverer) PreparePutManifest(ctx context.Context, host docker.RegistryHost, desc ocispec.Descriptor) (content.Writer, error) {
+	d.tracker.SetStatus(d.reference, docker.Status{
+		Status: content.Status{
+			Ref:       d.reference,
+			Total:     desc.Size,
+			Expected:  desc.Digest,
+			StartedAt: time.Now(),
+		},
+	})
+
+	req := d.request(host, http.MethodPut, "manifests", desc.Digest.String())
+	if req == nil {
+		return nil, errors.New("not implemented")
+	}
+
+	req.header.Set("Content-Type", desc.MediaType)
+
+	pr, pw := io.Pipe()
+	respC := make(chan response, 1)
+	body := ioutil.NopCloser(pr)
+
+	req.body = func() (io.ReadCloser, error) {
+		if body == nil {
+			return nil, errors.New("retry request, cannot reset the stream")
+		}
+
+		readout := body
+		body = nil
+		return readout, nil
+	}
+	req.size = desc.Size
+
+	go func() {
+		defer close(respC)
+		resp, err := req.doWithRetries(ctx, nil)
+		if err != nil {
+			respC <- response{err: err}
+			pr.CloseWithError(err)
+			return
+		}
+
+		switch resp.StatusCode {
+		case http.StatusOK, http.StatusCreated, http.StatusNoContent:
+		default:
+			err := remoteserrors.NewUnexpectedStatusErr(resp)
+			pr.CloseWithError(err)
+		}
+		respC <- response{Response: resp}
+	}()
+
+	return &artifactsManifest{
+		ref:       d.reference,
+		expected:  desc.Digest,
+		pipe:      pw,
+		responseC: respC,
+		tracker:   d.tracker,
+	}, nil
+}
+
+func (d *dockerDiscoverer) CheckManifest(ctx context.Context, host docker.RegistryHost, desc ocispec.Descriptor) error {
+	// Check if the manifest exists
+	req := d.request(host, http.MethodHead, "manifests", desc.Digest.String())
+	if req == nil {
+		return errors.New("not implemented")
+	}
+
+	req.header.Set("Accept", desc.MediaType)
+
+	resp, err := req.doWithRetries(ctx, nil)
+	if err != nil {
+		if !errors.Is(err, docker.ErrInvalidAuthorization) {
+			return err
+		}
+	} else {
+		if resp.StatusCode == http.StatusOK && resp.Header.Get("Docker-Content-Digest") == desc.Digest.String() {
+			d.tracker.SetStatus(d.reference, docker.Status{
+				Committed: true,
+				Status: content.Status{
+					Ref:    d.reference,
+					Total:  desc.Size,
+					Offset: desc.Size,
+				}})
+
+			return errdefs.ErrAlreadyExists
+		} else if resp.StatusCode != http.StatusNotFound {
+			err := remoteserrors.NewUnexpectedStatusErr(resp)
+			resp.Body.Close()
```

Comment:
Do you need to call Close in all code paths?

Reply:
good catch